### DISCUSSION TITLE
Filtering aggregates programatically

### DISCUSSION THREAD

**Discussion by yoopeyes** [eclipse-birt/birt#1102] [Q&A]
Hello,
I'm writing a BIRT report that calculates SUM values based on two parameters selected. I'm struggling to get it to work. I need it to calculate 'totalHomeValue' if parameter 'showhome' is ticked (and 'totalCurrencyValue' if it is not ticked).

I have three date selections (parameter called 'WhichDate'), given values 1 = PaymentDueDate, 2 = TaxDueDate & 3 = DocumentDueDate. I have fields called PaymentDueDateDays, TaxDueDateDays & DocumentDueDateDays. These are calculations made in SQL that show how many days before each of the payment dates are hit.

So, for example, an invoice might have a Payment due date of 3rd November, so the PaymentDueDateDays value would be 2, based on current date being the 1st. It might have a tax Date of 3rd December, so TaxDueDateDays would be 31.

I need the filter to only SUM the chosen value if the selected date type days value is less than 30. So, in effect, if WhichDate = 1, then sum all values where PaymentDueDateDays <= 30, if WhichDate = 2, then sum all values where TaxDueDateDays <= 30 and if WhichDate = 3, then sum all values where DocumentDueDateDays <= 30.

I've created an aggregate label. In the expression builder I've put:

if (params["ShowHome"].value = 1)
{
    row["totalHomeValue"];
}
else
{
    row["totalCurrencyValue"];
}
Then in the filter expression I've put:

switch(params["WhichDate"].value) {
    case 1:
        datasetrow["paymentduedatedays"] <= 30
        break;
    case 2:
        datasetrow["taxdatedays"] <= 30
        break;
    case 3:
        datasetrow["documentdays"] <= 30
        break;
}
It is calculating the values but not taking the filter conditions into account. I have another field that performs the calculation without any filtering and both are returning the same value. It suggests that the filtering condition isn't working properly. I've tried altering the code to just use if/else if statements instead of a switch statement but it then errors (and I don't know enough about BIRT to understand it).

So, given this code:

if (params["WhichDate"].value = 1)
{
row["paymentduedatedays"] <= 30;
}
else if
(params["WhichDate"].value = 2)
{
row["taxdatedays"] <= 30;
}
else
{
row["documentdatedays"]
}

the error is:

In nested aggregation "lessThan30", all the non-constant expression must have same group level.
It isn't immediately obvious to me what I'm doing wrong. Any help would be appreciated.

**Comment by hvbtup** [comment]
I didn't read and understand everything, but your JS code contains errors.
"=" is an assignment. Use "==" for equality testing.
